pkg/lib/bad/error.go

```diff
@@ -0,0 +1,184 @@
+// Package bad helps you deal with errors.
+package bad
+
+import (
+	"strings"
+)
+
+// ErrorSubject identifies where the error arose according to the component that
+// emitted it.
+type ErrorSubject string
+
+const (
+	// The error arose for an unspecified reason.
+	ErrSubjectNone ErrorSubject = ""
+	// The error arose because user input did not conform to the required
+	// specification. The user will need to modify their request before
+	// submitting it again.
+	ErrSubjectInput ErrorSubject = "input"
+	// The error arose because of an internal logic problem with the component
+	// that created the error, normally signaling a programming issue. The
+	// system developer will need to modify their code to meet this request.
+	ErrSubjectInternal ErrorSubject = "internal"
+	// The error arose because some other service or component that the
+	// component that created the error depends on also returned an error. The
+	// error may be transient, especially if this error is a leaf, or may
+	// require operator intervention (e.g. out of disk space).
+	ErrSubjectDependency ErrorSubject = "dependency"
+)
+
+// Error represents a node in a tree of errors. Error (and thus the whole tree)
+// is JSON marshallable so that a tree of Errors can be sent across HTTP.
+type Error struct {
+	// The type of the error. For native errors, this is just the error text,
```

Comment:
Ok so:

The `Type` field is intended to uniquely identify a type of error, and the type of what is in the `Data` field, so that the API Client can switch on the type field to work out how it wants to present the error – either by using the data in the `Data` field to work out/build up what it wants to display or just to replace the error message with it's own interpretation of what would be more helpful to the user. 

A good example of where this might happen is if there is some future browser-based UI that allows users to submit jobs, a validation error that comes back with a `Type` of e.g. `"Job did not pass validation"` and `Data` of 
```json
[
    {"field": "Name", "error": "contains null bytes"}, 
    {"field": "Docker.Image", "error": "image does not have a recognised platform", "validValues": ["linux/amd64", "linux/arm64", ...]}
]
``` 
In this case, the browser-based UI could recognise the `Type` and handle the error by highlighting all the form fields that are mentioned in the `Data` and can put the individual error messages next to those fields for the user to review in context.

Similarly, the CLI could do the same interpretation of the `Type` and show hints as I showed examples of in the design doc by working out where each field has taken it's value from and suggesting to the user how to modify these. So the UX is quite different for each client and the structured data allows each client to pick the best UX. 

So the requirement is that the `Type` should identify a way to **handle** this error uniquely. Different modules may want to emit the same `Type` if the way a client should handle this error is the same, or may want to emit different `Types` if the handling should be different. That's ultimately a case-by-case thing, and there's only so abstract that this can get – the granularity in `Types` needs to be driven by the actual clients people are developing and what logic people want to apply there. 

I deliberately chose against making `Type` an enum, to meet your requirement of modules being able to implement their own errors with little required visiblity at a higher level, which is sovling the existing problem with the bacerrors package requiring all error definitions to be centralised.

I am not a fan of error codes. It's just a level of indirection that I don't see as necessary – so my design is to just use a human-readable string as the type. With a code, ultimately anyone interpreting the code will need to look up its meaning in some dictionary, which is either a document we'll need to maintain or they'll have to go and read the code. At least with a human readable type string, the user can see the error for the first time coming from the API and immediately be able to understand it. The expectation is that `Type` field is a human-readable string that identifies the type of error, so doesn't contain specifics – think "job requested more resources than are available" than "job requested more memory (200G) than is available (14G)". 

Hopefully the above example also demonstrates why I didn't feel a `map[string]string` is appropriate – the level of detail that provides is not sufficient here, and [the example I show in the design doc](https://www.notion.so/expanso/Errors-b22b6b3ab35c4e12a92a4636b17d3c16?pvs=4#83072fa075644ac9b9e8a0df154648ed) is not a particularly complicated case but uses a lot of structure. Even the pre-existing errors [already require a level of detail beyond that](https://github.com/bacalhau-project/bacalhau/blob/9530b2bae9388465c679818af2ca359eeb8a9185/pkg/orchestrator/errors.go#L24-L28). I think the "future" you are talking about is already here – I make the case in the design doc as for why errors are not strings, and clients already need more detail to do any better interpretation of errors.

I picked `any` to allow modules to define their own data types that fit the specific situation, and there doesn't seem to be a more general interface. Can you say more about your concern with a more structured field? What is the risk with an arbitrary structure that makes it less maintainable? Yes, it will require us to document/define what things are going to come across the API and apply a level of rigour to not changing this interface arbitrarily... but we need to do that anyway, and even for a `map[string]string` it needs to have clear and slowly-changing keys with documentation on what format they take and what they mean, which is broadly the same information as for an arbitrary structure. 

I was assuming that `bad.Error` would at least be required to wrap native errors, because otherwise we'd be trying to send two different types as an error response (string or bad.Error), depending on which part of the code has been uplifted? In the design doc I talk about how for native errors [only outputting the final layer of the error message](https://www.notion.so/expanso/Errors-b22b6b3ab35c4e12a92a4636b17d3c16?pvs=4#b02ae231d2bc42f3a2ad148429631e1a) already beings immediate value without needing to visit every error site, so that's what I'm trying to achieve here.